### DISCUSSION TITLE
Help: websocket how to add heartbeat detection?

### DISCUSSION THREAD

**Discussion by wangziming133** [tokio-rs/axum#1127] [Q&A] [answered]
for example https://github.com/tokio-rs/axum/blob/main/examples/chat/src/main.rs

**Comment by davidpdrsn** [comment, marked as answer]
axum's websockets is built on tungstenite which afaik handles heartbeats automatically behind the scenes. If you wanna add something manual on top of that you can send a receive a message on an interval.

**Reply by BTOdell** [reply]
Just for additional context:

`tungstenite` only automatically responds to Ping messages from the client. It does not generate heartbeat messages from the server in order to keep the connection alive. This will need to be done manually on the server-side, or just require the client to initiate the heartbeat (ping) messages.

**Reply by LockedThread** [reply]
> Just for additional context:
> 
> `tungstenite` only automatically responds to Ping messages from the client. It does not generate heartbeat messages from the server in order to keep the connection alive. This will need to be done manually on the server-side, or just require the client to initiate the heartbeat (ping) messages.

Thank you for this.
